orttraining/orttraining/python/training/ortmodule/_utils.py

```diff
@@ -38,17 +39,74 @@ def _ortvalue_from_torch_tensor(torch_tensor):
     return C.OrtValue.from_dlpack(to_dlpack(torch_tensor), is_bool_tensor)
 
 
-def _torch_tensor_from_dl_pack(dlpack, ortvalue, device):
-    torch_tensor = from_dlpack(dlpack) if device.type != 'ort' else C.ort_from_dlpack(dlpack)
-    return torch_tensor.to(torch.bool) if ortvalue.data_type() == 'tensor(bool)' else torch_tensor
+def _ortvalues_to_torch_tensor(ortvalues, device):
+    if len(ortvalues) == 0:
+        return tuple()
+    if (hasattr(ortvalues, 'to_dlpack') and
+            all(ortvalue.proto_type() != 9 for ortvalue in ortvalues)):
+        if 'ort' == (device if isinstance(device, str) else device.type):
+            return tuple(ortvalues.to_dlpack(
+                lambda dlpack_structure: C.OrtValue.from_dlpack(dlpack_structure, False)))
+        return tuple(ortvalues.to_dlpack(_from_dlpack))
+
+    def _to_dlpack1(ortvalues, my_tensor):
+        tensors = []
+        for i in range(len(ortvalues)):
+            ov = ortvalues[i]
+            if hasattr(ov, '_ortvalue'):
+                ov = ov._ortvalue
+            dlp = ov.to_dlpack()
+            te = my_tensor(dlp)
+            tensors.append(te)
+        return tuple(tensors)
+
+    def _to_dlpack_at1(ortvalues, my_tensor):
+        tensors = []
+        for i in range(len(ortvalues)):
+            dlp = ortvalues.dlpack_at(i)
+            te = my_tensor(dlp)
+            tensors.append(te)
+        return tuple(tensors)
+
+    def _to_dlpack2(ortvalues, my_tensor):
+        tensors = []
+        for i in range(len(ortvalues)):
+            ov = ortvalues[i]
+            if hasattr(ov, '_ortvalue'):
+                ov = ov._ortvalue
+            dlp = ov.to_dlpack()
+            te = my_tensor(dlp, ov)
+            tensors.append(te)
+        return tuple(tensors)
+
+    def _to_dlpack_at2(ortvalues, my_tensor):
+        tensors = []
+        for i in range(len(ortvalues)):
+            dlp = ortvalues.dlpack_at(i)
+            te = my_tensor(dlp, ortvalues[i])
+            tensors.append(te)
+        return tuple(tensors)
+
+    if 'ort' == (device if isinstance(device, str) else device.type):
```

Comment:
I would like to clarify more on this. You are changing `_ortvalue_to_torch_tensor` to `_ortvalues_to_torch_tensor`. From the function name it means in any case we should return torch.Tensor (list) at the end. But in your new code, it returns an OrtValue list at line 48-49, and here, it return torch.Tensor list if `hasattr(C, 'ort_from_dlpack')` is true, but will still return OrtValue list if false at line 98-99. The main problem is `C.ort_from_dlpack` and `C.OrtValue.from_dlpack` are two different functions, one returns torch.Tensor while the other returns OrtValue. Just as Sherlock said, the names are comfusing, but they are different.

'ort' device type is a special device type, it's used to descibe the device of original OrtValue, just like 'cpu' and 'cuda', but not the target tensor type we need to generate. For this function, we need to always return torch.Tensor. 'ort' device is now only used in ort_eager mode. As I said previously, we don't have any UT or CI pipeline to cover this for now. if you want to try it, you need to use specific PyTorch version and build ORT using "--build_eager_mode" to enable it. You can run the test script at 'orttraining/orttraining/eager/test_model/mnist_fc_training.py' if you can have the ort_eager mode environment setup, but this is very hard to do for now. Acutally you should not use `if hasattr(C, 'ort_from_dlpack')` here to do different things. 'ort' device is for ort_eager mode, and ort_eager mode needs "--build_eager_mode", and when this is enabled, the `C` must have `ort_from_dlpack`. So the correct behavior here should be to `assert(hasattr(C, 'ort_from_dlpack'))` if device type is 'ort'. Your changes acutally break the ort_eager mode, which is used and developed by someone else.

In your new test code, I can see you added `isinstance(v, tensor_type)` but still pass `C_OrtValue` as the target type if device is 'ort': `self._ortvalues_to_torch_tensor_ortvaluevector('ort', C_OrtValue)`, which is logically wrong. As mentioned above, in any case, the returned tensor should be torch.Tensor. I think for any UT, no matter old one or newly added, we should run it on both the code before and after the changes. The new tests will definitely work on your new changed code, but you should also try to roll back your changes to old version and run the tests to see if they can pass, if not, the tests are not valid. But I guess you will not be able to do this for 'ort' device, as I mentioned above, it requires specific environment, which is not easy to setup, and normally we don't enable "--build_eager_mode" when building ORT. Any code using device 'ort' without "--build_eager_mode" is not valid.